---
title: "snpnet analysis"
author: "JKP"
date: "2023-02-13"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## R Markdown

First, install `snpnet` package according to instructions [here](https://github.com/junyangq/snpnet).

```{r snpnet, eval = FALSE}
library(devtools)

# First, install dependencies that are not on CRAN
install_github("junyangq/glmnetPlus", dependencies = TRUE)
## if this ^ does not work, run 
## curl -LO http://r.research.att.com/libs/gfortran-4.8.2-darwin13.tar.bz2
## sudo tar fvxj gfortran-4.8.2-darwin13.tar.bz2 -C /
# in the terminal and then try again.


install_github("chrchang/plink-ng", subdir="/2.0/cindex", dependencies = TRUE)
install_github("chrchang/plink-ng", subdir="/2.0/pgenlibr", dependencies = TRUE)

# Now install snpnet
install_github("junyangq/snpnet", dependencies = TRUE)
```

## Data

### Genomic data

Download the genomic data from [here](https://datadryad.org/stash/dataset/doi:10.5061/dryad.m0cfxpp10). We will use the full SNP data set, in the files `Genotypes.map` and `Genotypes.ped`.

We first prepare the data using PLINK v2 (I assume it is in your working folder). We do the standard quality control steps (remove SNPs with minor allele frequency lower than $0.01$, remove SNPs missing in more than $10\%$ of individuals and remove individuals missing more than $5\%$ of the SNPs). This might be unecessary because `snpnet` does this in-built. We also the dataset to the input format the `snpnet` needs.

```{r plink}
system2("./plink2", "--pedmap doi_10.5061_dryad.m0cfxpp10__v10/Genotypes --chr-set 32 --out sparrow_snps --maf 0.01 --geno 0.1 --mind 0.05 --make-pgen vzs")
```

### Phenotype data

```{r phenotype}
# Load the data
library(dplyr)
df <- read.table("doi_10.5061_dryad.m0cfxpp10__v10/Data.txt",
                 header = T)

# We should account fo
df2 <- select(df, c("age1wing", "age1tarsus", "age1mass"))

df2$IID = df$id
df2$FID <- 1
# Make the file that snpnet needs

write.csv(df2, file = "pheno.csv", row.names = F)

```

## Analysis

Do the analysis.

```{r analysis}
library(snpnet)

configs <- list(
    # results.dir = "PATH/TO/SAVE/DIR",  # needed when saving intermediate results
    # save = TRUE,  # save intermediate results per iteration (default FALSE)
    # nCores = 16,  # number of cores available (default 1)
    # niter = 100,  # max number of iterations (default 50)
    # prevIter = 15,  # if we want to start from some iteration saved in results.dir
    # use.glmnetPlus = TRUE,  # recommended for faster computation
    # early.stopping = FALSE,  # whether to stop based on validation performance (default TRUE)
    plink2.path = "./plink2",   # path to plink2 program
    zstdcat.path = "zstdcat"  # path to zstdcat program
)

# check if the provided paths are valid
for (name in names(configs)) {
    tryCatch(system(paste(configs[[name]], "-h"), ignore.stdout = T),
             condition = function(e) cat("Please add", configs[[name]], "to PATH, or modify the path in the configs list.")
    )
}

model <- snpnet(genotype.pfile = "sparrow_snps",
                phenotype.file = "pheno.csv",
                phenotype = "age1tarsus",
                covariates = NULL,
                # alpha = 1 means lasso, 0 means ridge, elastic net between
                alpha = 1, 
                nlambda = 100, 
                configs = configs)

```

